File: download-all-tracks.mjs
import * as fs from 'fs';
import { execSync } from 'child_process';

const scriptContent = fs.readFileSync('data.js', 'utf8');
// Evaluate DATA safely
const dataObjStr = scriptContent.replace('const DATA = {', 'global.DATA = {');
eval(dataObjStr);

const idsToDownload = new Set();

DATA.works.forEach(w => {
  if (w.categories && w.categories.includes('original')) {
    idsToDownload.add(w.id);
  }
});
if (DATA.bgVideos) DATA.bgVideos.forEach(id => idsToDownload.add(id));
if (DATA.videoThumbnails) DATA.videoThumbnails.forEach(id => idsToDownload.add(id));

const ids = Array.from(idsToDownload);
console.log(`Total unique tracks to ensure downloaded: ${ids.length}`);

for (const id of ids) {
    if (!fs.existsSync(`audio/${id}.webm`)) {
        console.log(`Downloading ${id}...`);
        try {
            execSync(`/Users/borjafernandezangulo/30_CORTEX/.venv/bin/yt-dlp -f 251 -o "audio/%(id)s.%(ext)s" "https://www.youtube.com/watch?v=${id}"`, {stdio: 'inherit'});
        } catch(e) { 
            console.error(`Failed ${id}`); 
        }
    } else {
        console.log(`${id} already exists, skipping.`);
    }
}

File: download_all_tracks.py
import subprocess
import os
import re

with open('data.js', 'r') as f:
    content = f.read()

ids_to_download = set()

vid_thumbs_match = re.search(r'videoThumbnails:\s*\[(.*?)\]', content, re.DOTALL)
if vid_thumbs_match:
    thumbs = re.findall(r"'([^']+)'", vid_thumbs_match.group(1))
    for t in thumbs:
        ids_to_download.add(t)

bg_vids_match = re.search(r'bgVideos:\s*\[(.*?)\]', content, re.DOTALL)
if bg_vids_match:
    bgs = re.findall(r"'([^']+)'", bg_vids_match.group(1))
    for b in bgs:
        ids_to_download.add(b)

# Extract objects in `works` array
lines = content.split('\n')
for line in lines:
    if 'id:' in line and 'categories:' in line:
        if '"original"' in line or "'original'" in line:
            id_match = re.search(r'id:\s*"([^"]+)"', line)
            if id_match:
                ids_to_download.add(id_match.group(1))

print(f"Total unique tracks to ensure downloaded: {len(ids_to_download)}")

for yt_id in ids_to_download:
    file_path = f"audio/{yt_id}.webm"
    if not os.path.exists(file_path):
        print(f"Downloading {yt_id}...")
        try:
            subprocess.run([
                '/Users/borjafernandezangulo/30_CORTEX/.venv/bin/yt-dlp', 
                '-f', '251', 
                '-o', f"audio/{yt_id}.%(ext)s", 
                f"https://www.youtube.com/watch?v={yt_id}"
            ], check=True)
        except subprocess.CalledProcessError:
            print(f"Failed {yt_id}")
    else:
        print(f"{yt_id} already exists, skipping.")
